### PR TITLE
[Source-MSSQL] Enable per-stream and Global states and migrate away from Legacy states

### PR DESCRIPTION
**Pull request by nguyenaiden**
### Description
When a customer syncs data from a DB source, it’s likely that they want to sync multiple tables, which means multiple streams in the recognized catalog. They want to be able to make changes to each stream and expect that each individual stream is an isolated entity. For examples, these are some common schema changes (non-exhaustive) that can happen:
- Adding/removing fields of a stream
- Adding/removing a stream from the catalog
- Updating a field type of a column in the DB
- Changing the cursor of a stream to another field

With the current global state implementation for Source-MSSQL, making any of these changes will require a connection-level reset that will affect the entire schema. 

As part of `source-mssql` bronze certification, this connector is being migrated to per-stream state to allow isolated changes at a stream level without having to reset the entire connection. Similar change for [MySQL](https://github.com/airbytehq/airbyte/pull/16007)

This change will both affect Standard and CDC syncs. Below are the before and after example states for CDC:

Before:
```
{
  "cdc": false,
  "streams": [
    {
      "stream_name": "purchases",
      "cursor_field": [
        "_ab_cdc_cursor"
      ],
      "stream_namespace": "dataset_1gb"
    }
  ],
  "cdc_state": {
    "state": {
      "mssql_cdc_offset": {
        "[\"mssql-benchmark-db\",{\"server\":\"mssql-benchmark-db\",\"database\":\"mssql-benchmark-db\"}]": "{\"commit_lsn\":\"0000da20:00019508:0004\",\"snapshot\":true,\"snapshot_completed\":true}"
      },
      "mssql_db_history": "[history data]"
    }
  }
}
```

After:
```
{
  "shared_state": {
    "state": {
      "mssql_cdc_offset": {
        "[\"mssql-benchmark-db\",{\"server\":\"mssql-benchmark-db\",\"database\":\"mssql-benchmark-db\"}]": "{\"transaction_id\":null,\"event_serial_no\":1,\"commit_lsn\":\"0000da20:000046c0:00cb\",\"change_lsn\":\"0000da20:00003dd8:0036\"}"
      },
      "mssql_db_history": "[history data]"
    }
  },
  "streamStates": [
    {
      "streamDescriptor": {
        "name": "purchases",
        "namespace": "dataset_1gb"
      },
      "streamState": {
        "stream_name": "purchases",
        "cursor_field": [
          "_ab_cdc_cursor"
        ],
        "stream_namespace": "dataset_1gb"
      }
    }
  ]
}
```
Note: This PR will be paired with an emergency rollback PR in case a customer upgrade goes wrong.

### CODE REVIEW

**Comment by akashkulk** [comment]
Can you just link the MySQL PR port here too?

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java [hunk 1]
```diff
@@ -267,7 +277,7 @@ public Set<String> getExcludedInternalNameSpaces() {
   public AirbyteCatalog discover(final JsonNode config) throws Exception {
     final AirbyteCatalog catalog = super.discover(config);
 
-    if (MssqlCdcHelper.isCdc(config)) {
+    if (isCdc(config)) {
```
Comment (stephane-airbyte) [inline airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java:280]: please don't do a static import of functions. It makes it a lot harder to know where the function lives without looking at the imports

Comment (nguyenaiden) [reply airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java:280]: Ack. I'll make the change.

**Review by akashkulk — COMMENTED** [review]
Nice! This seems as a small change (lot of test changes). The example you used here is for the GLOBAL state, does this also work for per-STREAM state?

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java [hunk 2]
```diff
@@ -102,6 +104,14 @@ public MssqlSource() {
     super(DRIVER_CLASS, AdaptiveStreamingQueryConfig::new, new MssqlSourceOperations());
   }
```
Comment (xiaohansong) [inline airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java:106]: maybe add a `@Override` annotation?

**Review by xiaohansong — COMMENTED** [review]
looks good to me!

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java
Comment (akashkulk) [inline airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java:1]: As discussed offline, we need to handle the case in CDC mode where new streams are added and need to be snapshotted. With the current implementation, we'll continue reading off of the WAL log so will get subsequent changes but never the initial snapshots. 

You can use
- AbstractJdbcSource.identifyStreamsToSnapshot
- AirbyteDebeziumHandler.getSnapshotIterators

to achieve this

Comment (nguyenaiden) [reply airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java:1]: Ack, handled with latest push.

File: docs/integrations/sources/mssql.md [hunk 3]
```diff
@@ -342,6 +342,7 @@ WHERE actor_definition_id ='b5ea17b1-f170-46dc-bc31-cc744ca984c1' AND (configura
 
 | Version | Date       | Pull Request                                                                                                      | Subject                                                                                                                                         |
 |:--------|:-----------|:------------------------------------------------------------------------------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------|
+| 3.1.1   | 2023-11-30 | [33018](https://github.com/airbytehq/airbyte/pull/33018)                                                          | Migrate to Per-stream/Global states and away from Legacy states                                                                                 |
```
Comment (akashkulk) [inline docs/integrations/sources/mssql.md:345]: Let's make this a minor bump (3.2.0)

**Review by akashkulk — APPROVED** [review]

**Comment by nguyenaiden** [comment]
@akashkulk I'm going to push up the change that includes the test for per stream before merging.

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java [hunk 4]
```diff
@@ -490,18 +499,45 @@ public List<AutoCloseableIterator<AirbyteMessage>> getIncrementalIterators(
...
           mssqlCdcConnectorMetadataInjector,
-          MssqlCdcHelper.getDebeziumProperties(database, catalog),
+          MssqlCdcHelper.getDebeziumProperties(database, catalog, false),
           DebeziumPropertiesManager.DebeziumConnectorType.RELATIONALDB,
-          emittedAt, true);
+          emittedAt,
+          true);
+
+      /*
+       * If the CDC state is null or there is no streams to snapshot, that means no stream has gone
```
Comment (rodireich) [inline airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java:519]: Is this correct? if no stream finished snapshot we want to snapshot all, not have an incremental sync, no?

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java [hunk 5]
```diff
@@ -490,18 +499,45 @@ public List<AutoCloseableIterator<AirbyteMessage>> getIncrementalIterators(
...
+              mssqlCdcConnectorMetadataInjector,
+              MssqlCdcHelper.getDebeziumProperties(database, catalog, true),
+              new MssqlCdcStateHandler(stateManager),
+              DebeziumPropertiesManager.DebeziumConnectorType.RELATIONALDB,
+              emittedAt);
+      /*
+       * The incremental iterators needs to be wrapped in a lazy iterator since only 1 Debezium engine for
+       * the DB can be running at a time
+       */
+      return List.of(snapshotIterators, AutoCloseableIterators.lazyIterator(incrementalIteratorsSupplier, null));
```
Comment (rodireich) [inline airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java:540]: why `snapshotIterators` not included in lazy iterator?

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java (hunk 4 above)
Comment (nguyenaiden) [reply airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java:519]: The naming is incorrect. It's actually snapshot for _new_streams.

**Pull request merged**